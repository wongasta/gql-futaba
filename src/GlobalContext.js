import React from "react";

export const GlobalContext = React.createContext({
  user: null,
  user_id: null
});