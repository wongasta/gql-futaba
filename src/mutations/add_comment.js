//@flow
import {commitMutation} from 'react-relay';
import graphql from 'babel-plugin-relay/macro';
import type {Environment as EnvironmentType} from 'relay-runtime';

export type PostInputType={
  user: ?string,
  post: string,
  comment_content: string,
  image_url: ?string
}
export type CommentType={
    id: string,
    user: string,
    user_id: string,
    post: string,
    title: void,
    comment_content: string,
    image_url: ?string,
    created_ts: number
}
export type CommentEdgeType={
  cursor: string,
  node: CommentType
}
type AddCommentPayloadType={
  commentEdge: CommentEdgeType
}
type SuccessCbType=(AddCommentPayloadType)=>void

const mutation = graphql`
  mutation addCommentMutation($input: addCommentInput!){
    add_comment(input: $input){
      commentEdge{
        cursor
        node{
          id
          user
          user_id
          post
          comment_content
          image_url
          created_ts
        }
      }
    }
  }
`;

export default (function add_comment(environment, postInput, successCb){
  commitMutation(
    environment,
    {
      mutation,
      variables: {
        input: postInput
      },
      onCompleted: (response, error)=>{
        successCb(response.add_comment);
      },
      onError: (error)=>{
        console.log(error)
      },
      updater: (store, data)=>{
        const newCommentEdge = store.getRootField('add_comment')?.getLinkedRecord('commentEdge');
        const comments = store.get(postInput.post)?.getLinkedRecord('comments',{first: 2147483647});
        comments?.setLinkedRecords([...(comments?.getLinkedRecords('edges') || []), newCommentEdge], 'edges');
      }
    }
  )
}: (EnvironmentType,PostInputType,SuccessCbType)=>void)