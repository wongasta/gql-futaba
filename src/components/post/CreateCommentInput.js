import React, {useRef,useContext} from "react";
import '../form_shared.css';
import {GlobalContext} from "../../GlobalContext";
import add_comment from "../../mutations/add_comment";

export default function CreateCommentInput({postId,environment}){
  const commentForm = useRef(null);
  const {user,user_id} = useContext(GlobalContext);

  function handleSubmit(e){
    const form=commentForm.current;
    const PostInput={
      user: user,
      post: postId,
      comment_content: form.querySelector('#input_content').value,
      image_url: form.querySelector('#input_image_url').value
    }
    add_comment(environment,PostInput,(data)=>{
      form.querySelector('#input_content').value="";
      form.querySelector('#input_image_url').value="";
    });
    e.preventDefault();
  }

  return (
    <form id="comment_form" className={"form_container"} ref={commentForm} onSubmit={handleSubmit}>
      <div className={"input_container"}>
        <div className={"labels"}>
          <label htmlFor="input_user">User</label>
          <label htmlFor="input_content" className="content">Comment</label>
          <label htmlFor="input_image_url">Image</label>
        </div>
        <div className={"inputs"}>
          <input id="input_user" type="text" value={user_id} disabled={true} />
          <textarea id="input_content" defaultValue={"Reply"}></textarea>
          <input id="input_image_url" type="url" placeholder="Image Link" />
        </div>
      </div>
      <div className={"submit_container"}>
        <button type="submit">Submit</button>
      </div>
    </form>
  )
}