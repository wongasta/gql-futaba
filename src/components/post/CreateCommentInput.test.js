import React from 'react';
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import CreateCommentInput from './CreateCommentInput';
import {GlobalContext} from "../../GlobalContext";
import add_comment from "../../mutations/add_comment";
jest.mock('../../mutations/add_comment', ()=>({
  __esModule: true,
  default: jest.fn()
}));

describe("CreateCommentInput", ()=>{

  afterAll(()=>{
    jest.clearAllMocks();
  })

  test("Construction Snapshot", ()=>{
    const Component = mount(<GlobalContext.Provider value={{user: "abc", user_id: "abc"}}><CreateCommentInput postId={"abc"} /></GlobalContext.Provider>);
    expect(toJson(Component)).toMatchSnapshot();
  });

  test("Submission", ()=>{
    add_comment.mockImplementation((_,__,cb)=>cb());
    const Component = mount(<GlobalContext.Provider value={{user: "abc", user_id: "abc"}}><CreateCommentInput postId={"abc"} /></GlobalContext.Provider>);
    Component.find("#input_content").instance().value="mock content";
    Component.find("#input_image_url").instance().value="mock image url";
    Component.update();
    Component.find("#comment_form").simulate('submit');
    expect(add_comment).toHaveBeenCalledTimes(1);
    expect(add_comment.mock.calls[0][1].user).toBe('abc');
    expect(add_comment.mock.calls[0][1].post).toBe('abc');
    expect(add_comment.mock.calls[0][1].comment_content).toBe('mock content');
    expect(add_comment.mock.calls[0][1].image_url).toBe('mock image url');
    expect(Component.find("#input_content").instance().value).toBe("");
    expect(Component.find("#input_image_url").instance().value).toBe("");
  });
})