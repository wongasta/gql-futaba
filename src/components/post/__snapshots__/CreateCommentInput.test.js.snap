// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCommentInput Construction Snapshot 1`] = `
<CreateCommentInput
  postId="abc"
>
  <form
    className="form_container"
    id="comment_form"
    onSubmit={[Function]}
  >
    <div
      className="input_container"
    >
      <div
        className="labels"
      >
        <label
          htmlFor="input_user"
        >
          User
        </label>
        <label
          className="content"
          htmlFor="input_content"
        >
          Comment
        </label>
        <label
          htmlFor="input_image_url"
        >
          Image
        </label>
      </div>
      <div
        className="inputs"
      >
        <input
          disabled={true}
          id="input_user"
          type="text"
          value="abc"
        />
        <textarea
          defaultValue="Reply"
          id="input_content"
        />
        <input
          id="input_image_url"
          placeholder="Image Link"
          type="url"
        />
      </div>
    </div>
    <div
      className="submit_container"
    >
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</CreateCommentInput>
`;
